### PR TITLE
Fix for #1124 issue added verifyThenAssertThat(Duration)

### PR DESCRIPTION
**Pull request by Alex-Diez**
Added `verifyThenAssertThat` method with custom `Duration` as parameter to `reactor.test.StepVerifier`

@smaldini can you please review?

I have submitted ICLA

### CODE REVIEW

File: reactor-test/src/main/java/reactor/test/StepVerifier.java [hunk 1]
```diff
@@ -322,6 +322,25 @@ static void resetDefaultTimeout() {
...
+	 * the stream has been terminated (either through {@link Subscriber#onComplete()},
+	 * {@link Subscriber#onError(Throwable)} or {@link Subscription#cancel()}).
+	 * Depending on the declared expectations and actions, notably in case of undersized
+	 * manual requests, such a verification could also block indefinitely. Use
+	 * {@link Duration} {@code duration} parameter to add a timeout on verify()-derived
+	 * methods.
+	 *
+	 * @param duration the maximum duration to wait for the sequence to terminate, or
+	 * {@link Duration#ZERO} for unlimited wait.
+	 * @return the actual {@link Duration} the verification took.
```
Comment (simonbasle) [inline reactor-test/src/main/java/reactor/test/StepVerifier.java]: the `@return` and `@throws` blocks are incorrect (copy-paste).

File: reactor-test/src/main/java/reactor/test/StepVerifier.java [hunk 2]
```diff
@@ -322,6 +322,25 @@ static void resetDefaultTimeout() {
...
+	/**
+	 * {@link #verify() Verifies} the signals received by this subscriber, then exposes
+	 * various {@link Assertions assertion methods} on the final state.
+	 * <p>
+	 * Note that like {@link #verify()}, this method will <strong>block</strong> until
+	 * the stream has been terminated (either through {@link Subscriber#onComplete()},
+	 * {@link Subscriber#onError(Throwable)} or {@link Subscription#cancel()}).
+	 * Depending on the declared expectations and actions, notably in case of undersized
+	 * manual requests, such a verification could also block indefinitely. Use
+	 * {@link Duration} {@code duration} parameter to add a timeout on verify()-derived
```
Comment (simonbasle) [inline reactor-test/src/main/java/reactor/test/StepVerifier.java]: I'd rephrase the sentence to "`As a consequence you can use the {@link Duration} {@code duration} parameter to set a timeout.`".

File: reactor-test/src/test/java/reactor/test/StepVerifierTimeoutTests.java [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+				.isThrownBy(() ->
+						StepVerifier.create(Mono.delay(Duration.ofMillis(150)))
+						            .expectComplete()
+						            .verifyThenAssertThat(Duration.ofMillis(50)))
+				.withMessageStartingWith("VerifySubscriber timed out");
+	}
+
+	@Test
+	public void verifyThenAssertUsesCustomTimeout() {
+		StepVerifier.create(Mono.delay(Duration.ofMillis(150)))
```
Comment (simonbasle) [inline reactor-test/src/test/java/reactor/test/StepVerifierTimeoutTests.java]: maybe set the "default verify timeout" to 50ms (and reset it in a `finally` block)?
=> `StepVerifier.setDefaultTimeout(Duration.ofMillis(50))`

File: reactor-test/src/main/java/reactor/test/DefaultStepVerifierBuilder.java [hunk 4]
```diff
@@ -79,6 +79,51 @@
 	 */
 	static Duration defaultVerifyTimeout = StepVerifier.DEFAULT_VERIFY_TIMEOUT;
 
+	private static class StreamStepRecorder {
```
Comment (simonbasle) [inline reactor-test/src/main/java/reactor/test/DefaultStepVerifierBuilder.java]: Can you rename to `HookRecorder`? (even if it is `private`, I think it is more fitting than `StreamStep`).

File: reactor-test/src/main/java/reactor/test/DefaultStepVerifierBuilder.java [hunk 5]
```diff
@@ -79,6 +79,51 @@
...
+		final Queue<Object>                                   droppedElements = new ConcurrentLinkedQueue<>();
+		final Queue<Throwable>                                droppedErrors   = new ConcurrentLinkedQueue<>();
+		final Queue<Tuple2<Optional<Throwable>, Optional<?>>> operatorErrors  = new ConcurrentLinkedQueue<>();
+
+		public void pluginHooks() {
```
Comment (simonbasle) [inline reactor-test/src/main/java/reactor/test/DefaultStepVerifierBuilder.java]: can you rename to `plugHooks()`? (nitpicking)

File: reactor-test/src/main/java/reactor/test/DefaultStepVerifierBuilder.java [hunk 6]
```diff
@@ -79,6 +79,51 @@
...
+			Hooks.onErrorDropped(droppedErrors::offer);
+			Hooks.onNextDropped(droppedElements::offer);
+			Hooks.onOperatorError((t, d) -> {
+				operatorErrors.offer(Tuples.of(Optional.ofNullable(t), Optional.ofNullable(d)));
+				return t;
+			});
+		}
+
+		public void unpluginHooks() {
```
Comment (simonbasle) [inline reactor-test/src/main/java/reactor/test/DefaultStepVerifierBuilder.java]: can you rename to `unplugHooks()`? (nitpicking)

**Comment by Alex-Diez** [comment]
@simonbasle, review comments addressed.

**Comment by simonbasle** [comment]
Thanks for the 🎁of PR @Alex-Diez, merged 👍
